Separate type for unaggregated network attestations

rkapka opened the pull request:
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Implements https://github.com/ethereum/consensus-specs/pull/3900

New changes after initial reviews
--

- I created new signed and unsigned `AggregateAndProof` types that hold `SingleAttestation`, which is needed to be able to save single attestations into pending atts, and simplified `GetCommitteeIndex` so that it doesn't return an error. Returning errors on getters is always very annoying to deal with, and we shouldn't get into the error scenario in the first place (most, if not all, potential errors are already handled in gossip checks).
- I save the new `SignedAggregateAttestationAndProofSingle` type to pending atts and convert the underlying attestation into an `AttestationElectra` when processing this attestation. This is after `validateUnaggregatedAttWithState`.
`validateUnaggregatedAttWithState` returns the committee so that we can convert the attestation.
- I removed the `validateBitLength` helper because I need the committee outside of this function too, and it's already available before the function call. So calling `VerifyBitfieldLength` directly is simpler.
- ~~I renamed `validateCommitteeIndexAndCount` to `validateCommitteeIndex` as that's how it was called before Electra changes. We don't need the Electra `validateCommitteeIndex` and `validateCommitteeIndexElectra` functions because their main purpose was to validate the Electra version, and it's no longer needed with single attestation (there is `validateAttestingIndex` instead)~~

Original description
--

The above consensus change introduces a new `SingleAttestation` type for unaggregated attestations. `AttestationElectra` will still be used for aggregated atts. Using the single version in all of our code would be a huge undertaking because many components, such as the attestation pool and attestation caches, require attestations to be of the old format. Modifying these components to work both with pre-Electra attestations and the single attestation is risky and too much work for Pectra.

The main reason why `SingleAttestation` was introduced is to prevent a DoS attack, which means it's main value is at the level of gossip. Once we pass gossip validation checks, there is no downside to using the `AttestationElectra` for unaggregated attestations too. This allows us to leave most attestation-related code untouched. To have the best of both worlds, we convert the single attestation to an Electra attestation as quickly as possible and use the latter from that point on.

When converting from `SingleAttestation` to `ElectraAttestation` we need to find the index _within the committee_ of the `attester_index` from the single attestation, because `attester_index` is the validator index within the whole state. To give an example:
> Validator 9999 is an attester and the committee index to which it is assigned is 5. Within committee number 5, the attester's position/index is 7. `SingleAttestation`'s values will be `attester_index = 9999` and `committee_index = 5`. When converting to an `ElectraAttestation`, we need to set bit number 7 in `attestation_bits`. We need to obtain the number 7 somehow as it is not present in the `SingleAttestation`.

The way it's done is by using the `AttestationStateFetcher` to get the attestation's target state, then get the appropriate committee from that state and lastly call `ToAttestationElectra` on the `SingleAttestation`, passing the committee to the function.

**Other notes for review**

- Obtaining the target state, getting the committee and converting takes less than 1ms. I presume it's because both the state and committee are cached.
- Tested on Kurtosis with Prysm-only nodes and and Prysm+Lodestar.
- The same "convert quickly to Electra attestation" design is what other clients are doing too to minimize disruption.

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have made an appropriate entry to [CHANGELOG.md](https://github.com/prysmaticlabs/prysm/blob/develop/CHANGELOG.md).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

terencechain commented on beacon-chain/sync/validate_beacon_attestation.go:
```diff
@@ -71,6 +70,32 @@ func (s *Service) validateCommitteeIndexBeaconAttestation(ctx context.Context, p
+		if !ok {
+			return pubsub.ValidationIgnore, fmt.Errorf("attestation has wrong type (expected %T, got %T)", &eth.SingleAttestation{}, att)
+		}
+		att = singleAtt.ToAttestationElectra(committee)
```
This is the biggest additional performance penalty as far as I can see. For a node that subscribe to several subnets. Is this a concern? We will call this 10-30k times per slot

nisdas commented on beacon-chain/sync/validate_beacon_attestation.go:
```diff
@@ -71,6 +70,32 @@ func (s *Service) validateCommitteeIndexBeaconAttestation(ctx context.Context, p
 		return pubsub.ValidationIgnore, nil
 	}
+
+	preState, err := s.cfg.chain.AttestationTargetState(ctx, data.Target)
```
Why are we calling this here ? There must be strong justification for 
- fetching a state
- fetching the committee from it

As these are additional bottlenecks on our highest traffic gossip path

nisdas replied:
Also why is this being called before the current gossip checks

rkapka commented on beacon-chain/sync/validate_aggregate_proof_test.go:
```diff
@@ -94,13 +94,6 @@ func TestVerifyIndexInCommittee_ExistsInBeaconCommittee(t *testing.T) {
 	assert.ErrorContains(t, wanted, err)
 	assert.Equal(t, pubsub.ValidationReject, result)
 
-	att.AggregationBits = bitfield.NewBitlist(1)
-	committeeIndex, err := att.GetCommitteeIndex()
-	require.NoError(t, err)
-	_, result, err = service.validateBitLength(ctx, s, att.Data.Slot, committeeIndex, att.AggregationBits)
-	require.ErrorContains(t, "wanted participants bitfield length 4, got: 1", err)
-	assert.Equal(t, pubsub.ValidationReject, result)
-
```
Because I removed the helper function. We still test this, but it's done in https://github.com/prysmaticlabs/prysm/pull/14659/files#diff-f1d6b0b8f81c14d3cf2ef27e17d4c4e7e9b1ad1f8fc3b5a66d9bdf2eeb615853 now.

rkapka replied:
never mind, I will revert this change

prestonvanloon replied:
> No, because of "I removed the validateBitLength helper because I need the committee outside of this function too, and it's already available before the function call. So calling VerifyBitfieldLength directly is simpler."

https://github.com/prysmaticlabs/prysm/pull/14659#issuecomment-2575296082

nisdas commented on beacon-chain/sync/validate_aggregate_proof.go:
```diff
@@ -175,10 +175,14 @@ func (s *Service) validateAggregatedAtt(ctx context.Context, signed ethpb.Signed
+		return pubsub.ValidationIgnore, err
+	}
 
-	committee, result, err := s.validateBitLength(ctx, bs, aggregate.GetData().Slot, committeeIndex, aggregate.GetAggregationBits())
```
why was this removed ? Its effectively the same as what has been added in

rkapka replied:
`validateBitLength` extracted the comittee from state but I need it outside for single attestation. I did this to prevent reading it twice

rkapka replied:
you're right, I will change it back

prestonvanloon replied:
Are you still planning to revert this change?

prestonvanloon replied:
> No, because of "I removed the validateBitLength helper because I need the committee outside of this function too, and it's already available before the function call. So calling VerifyBitfieldLength directly is simpler."


https://github.com/prysmaticlabs/prysm/pull/14659#issuecomment-2575296082

nisdas commented on beacon-chain/sync/validate_beacon_attestation.go:
```diff
@@ -259,12 +281,15 @@ func (s *Service) validateUnaggregatedAttWithState(ctx context.Context, a eth.At
-	committee, result, err := s.validateBitLength(ctx, bs, a.GetData().Slot, committeeIndex, a.GetAggregationBits())
-	if result != pubsub.ValidationAccept {
-		return result, err
+	committee, err := helpers.BeaconCommitteeFromState(ctx, bs, a.GetData().Slot, committeeIndex)
```
Same here, why is `validateBitLength` removed

nisdas commented on beacon-chain/sync/validate_beacon_attestation.go:
```diff
@@ -115,6 +104,48 @@ func (s *Service) validateCommitteeIndexBeaconAttestation(ctx context.Context, p
 		}
 	}
 
+	if !s.cfg.chain.InForkchoice(bytesutil.ToBytes32(data.BeaconBlockRoot)) {
```
Executing all these methods before we have validated `data.BeaconBlockRoot` is wrong. As we might have not processed the block yet (and it has to be inserted into our pending queue)

nisdas commented on beacon-chain/sync/validate_beacon_attestation.go:
```diff
@@ -115,6 +104,48 @@ func (s *Service) validateCommitteeIndexBeaconAttestation(ctx context.Context, p
+		return pubsub.ValidationIgnore, blockchain.ErrNotDescendantOfFinalized
+	}
+
+	if err = s.cfg.chain.VerifyLmdFfgConsistency(ctx, att); err != nil {
```
Same here and for the rest of the methods that follow it. If it is difficult enough, you should just stream the attestations after

rkapka commented on validator/client/beacon-api/propose_attestation_test.go:
```diff
@@ -258,86 +248,41 @@ func TestProposeAttestationElectra(t *testing.T) {
-			expectedErrorMessage: "attestation's bitfield can't be nil",
-		},
-		{
-			name: "nil signature",
-			attestation: &ethpb.AttestationElectra{
-				AggregationBits: testhelpers.FillByteSlice(4, 74),
-				Data: &ethpb.AttestationData{
-					Source: &ethpb.Checkpoint{},
-					Target: &ethpb.Checkpoint{},
-				},
-				CommitteeBits: testhelpers.FillByteSlice(8, 83),
-			},
-			expectedErrorMessage: "attestation signature can't be nil",
```
I removed it because we had two functions in the codebase checking if an attestation is nil. One is `helpers.ValidateNilAttestation` and the other one is `validateNilAttestation` on the VC side. The only difference between them was signature validation which is missing in the `helpers` package. I decided it's better to just have a single implementation and I expect the one in `helpers` to be more accurate. I am not exactly sure what happens though when the BN receives an attestation without a signature if the function in `helpers` doesn't fail.

potuz commented on beacon-chain/sync/validate_beacon_attestation.go:
```diff
@@ -141,23 +162,25 @@ func (s *Service) validateCommitteeIndexBeaconAttestation(ctx context.Context, p
 		tracing.AnnotateError(span, blockchain.ErrNotDescendantOfFinalized)
 		return pubsub.ValidationIgnore, blockchain.ErrNotDescendantOfFinalized
 	}
-	if err := s.cfg.chain.VerifyLmdFfgConsistency(ctx, att); err != nil {
```
This check is performed now after you are grabbing a full state and computing a beacon committee shuffling, this opens up the door for a malicious peer to kill the node with a crafted attestation.

terencechain commented on beacon-chain/rpc/prysm/v1alpha1/validator/attester.go:
```diff
@@ -62,24 +63,28 @@ func (vs *Server) ProposeAttestation(ctx context.Context, att *ethpb.Attestation
 	defer span.End()
 
-	committeeIndex, err := att.GetCommitteeIndex()
+	targetState, err := vs.AttestationStateFetcher.AttestationTargetState(ctx, att.Data.Target)
```
Any concerns with this line: for every attestation we propose, we need to retrieve the target state. What are the implications of this?

prestonvanloon replied:
Why not just return the committee? This is what you need anyway instead of the full state.

rkapka commented on beacon-chain/rpc/eth/beacon/handlers_pool.go:
```diff
@@ -335,28 +347,20 @@ func (s *Server) handleAttestationsElectra(ctx context.Context, data json.RawMes
+			if err = s.AttestationCache.Add(att.ToAttestationElectra(committee)); err != nil {
 				log.WithError(err).Error("could not save attestation")
 			}
-		} else if att.IsAggregated() {
```
single attestation can't be aggregated

rkapka commented:
New changes
--

- I created new signed and unsigned `AggregateAndProof` types and simplified `GetCommitteeIndex` so that it doesn't return an error. Returning errors on getters is always very annoying to deal with, and we shouldn't get into the error scenario in the first place (most, if not all, potential errors are already handled in gossip checks).
- I save the new `SignedAggregateAttestationAndProofSingle` type to pending atts and convert the underlying attestation into an `AttestationElectra` when processing this attestation. This is after `validateUnaggregatedAttWithState`.
validateUnaggregatedAttWithState returns the committee so that we can convert the attestation.
- I removed the `validateBitLength` helper because I need the committee outside of this function too, and it's already available before the function call. So calling `VerifyBitfieldLength` directly is simpler.
- I renamed `validateCommitteeIndexAndCount` to `validateCommitteeIndex` as that's how it was called before Electra changes. We don't need the Electra `validateCommitteeIndex` and `validateCommitteeIndexElectra` functions because their main purpose was to validate the Electra version, and it's no longer needed with single attestation (there is `validateAttestingIndex` instead)

rkapka commented on beacon-chain/sync/pending_attestations_queue_test.go:
```diff
@@ -92,18 +92,9 @@ func TestProcessPendingAtts_HasBlockSaveUnAggregatedAtt(t *testing.T) {
 		att.Signature = privKeys[i].Sign(hashTreeRoot[:]).Marshal()
 	}
 
-	// Arbitrary aggregator index for testing purposes.
```
This part is not needed for unaggregated atts

prestonvanloon reviewed (commented):
Partial review. Still looking at sync package

prestonvanloon commented on beacon-chain/sync/validate_beacon_attestation.go:
```diff
@@ -222,83 +236,93 @@ func (s *Service) validateUnaggregatedAttTopic(ctx context.Context, a eth.Att, b
 }
 
-func (s *Service) validateCommitteeIndexAndCount(
+func (s *Service) validateCommitteeIndex(
```
This method has been renamed, but it is still validating the committee count. Want to revert the renaming or refactor further?

rkapka replied:
You're right, I will change it back

potuz commented on beacon-chain/rpc/eth/beacon/handlers_pool.go:
```diff
@@ -317,14 +320,23 @@ func (s *Server) handleAttestationsElectra(ctx context.Context, data json.RawMes
 		// Broadcast the unaggregated attestation on a feed to notify other services in the beacon node
 		// of a received unaggregated attestation.
 		// Note we can't send for aggregated att because we don't have selection proof.
 		if !att.IsAggregated() {
```
This check is useless, the structure `SingleAttestation` always returns `false` here.

potuz commented on beacon-chain/rpc/eth/beacon/handlers_pool.go:
```diff
@@ -335,28 +347,20 @@ func (s *Server) handleAttestationsElectra(ctx context.Context, data json.RawMes
 
 		if features.Get().EnableExperimentalAttestationPool {
-			if err = s.AttestationCache.Add(att); err != nil {
+			if err = s.AttestationCache.Add(att.ToAttestationElectra(committee)); err != nil {
```
The conversion to `ToAttestationElectra` happens unconditionally above, I would keep this object and reuse it here.

potuz commented on beacon-chain/rpc/eth/beacon/handlers_pool.go:
```diff
@@ -335,28 +347,20 @@ func (s *Server) handleAttestationsElectra(ctx context.Context, data json.RawMes
 		} else {
-			if err = s.AttestationsPool.SaveUnaggregatedAttestation(att); err != nil {
-				log.WithError(err).Error("could not save unaggregated attestation")
+			if err = s.AttestationsPool.SaveUnaggregatedAttestation(att.ToAttestationElectra(committee)); err != nil {
```
Same here.

rkapka commented on beacon-chain/sync/validate_beacon_attestation.go:
```diff
@@ -71,14 +70,6 @@ func (s *Service) validateCommitteeIndexBeaconAttestation(ctx context.Context, p
 	if data.Slot == 0 {
 		return pubsub.ValidationIgnore, nil
 	}
-	// Broadcast the unaggregated attestation on a feed to notify other services in the beacon node
-	// of a received unaggregated attestation.
-	s.cfg.attestationNotifier.OperationFeed().Send(&feed.Event{
-		Type: operation.UnaggregatedAttReceived,
-		Data: &operation.UnAggregatedAttReceivedData{
-			Attestation: att,
-		},
-	})
```
This is not removed, it's now near the end of the function

rkapka commented:
@prestonvanloon 

> Do you still need to revert this change?

No, because of "I removed the validateBitLength helper because I need the committee outside of this function too, and it's already available before the function call. So calling VerifyBitfieldLength directly is simpler."

> Are you still planning to revert this change?

No, see above

> Why not just return the committee? This is what you need anyway instead of the full state.

To get the committee, we need to call `helpers.BeaconCommitteeFromState` and this function needs the state. Committees are not part of the state itself, they are computed on request.

potuz commented on beacon-chain/rpc/eth/validator/handlers.go:
```diff
@@ -205,11 +205,7 @@ func matchingAtts(atts []ethpbalpha.Att, slot primitives.Slot, attDataRoot []byt
 		// compare the committee index separately.
 		if postElectra {
 			if att.Version() >= version.Electra {
```
The whole block reads better as follows

```
        root, err := att.GetData().HashTreeRoot()
		if err != nil {
			return nil, errors.Wrap(err, "could not get attestation data root")
		}
		if !bytes.Equal(root[:], attDataRoot) {
			continue
		}
        if att.Version() < version.Electra || (postElectra && att.GetCommitteeIndex()  == index) {
			result = append(result, att)
        }
```

rkapka replied:
Agreed, but we need to have `(!postElectra && att.Version() < version.Electra) || ...`

potuz replied:
that should be covered for att.Version() < version.Electra since we should clean the pool at the fork, but anyway you can change the first side of the `||` to include the check
```
        if (!postElectra && att.Version() < version.Electra) || (postElectra && att.GetCommitteeIndex()  == index) {
```

potuz reviewed (commented):
First pass, will continue in 30 minutes

terencechain commented on api/server/structs/other.go:
```diff
@@ -36,6 +36,13 @@ type AttestationElectra struct {
 	CommitteeBits   string           `json:"committee_bits"`
 }
 
+type SingleAttestation struct {
```
I think `AttestationElectra` is still part of beacon block so it's needed for block related api end points right?

terencechain commented on beacon-chain/rpc/eth/beacon/handlers_pool.go:
```diff
@@ -335,28 +347,20 @@ func (s *Server) handleAttestationsElectra(ctx context.Context, data json.RawMes
-			return nil, nil, errors.Wrap(err, "failed to retrieve attestation committee index")
-		}
-		subnet := corehelpers.ComputeSubnetFromCommitteeAndSlot(uint64(len(vals)), committeeIndex, att.Data.Slot)
+		subnet := corehelpers.ComputeSubnetFromCommitteeAndSlot(uint64(len(vals)), att.GetCommitteeIndex(), att.Data.Slot)
```
Unrelated to this PR but calling `s.HeadFetcher.HeadValidatorsIndices(ctx, wantedEpoch)` allocate a new slice of active validator indices, it's expensive to just get the active validator count. `ActiveValidatorCount` is more suitable for this

rkapka replied:
I will add this comment to my follow-up list

terencechain commented on beacon-chain/rpc/prysm/v1alpha1/validator/attester.go:
```diff
@@ -69,29 +70,34 @@ func (vs *Server) ProposeAttestation(ctx context.Context, att *ethpb.Attestation
 
 	if features.Get().EnableExperimentalAttestationPool {
-		if err = vs.AttestationCache.Add(att); err != nil {
+		if err = vs.AttestationCache.Add(att.ToAttestationElectra(committee)); err != nil {
```
may be able to reuse `att.ToAttestationElectra(committee)` here

terencechain commented on beacon-chain/rpc/prysm/v1alpha1/validator/attester.go:
```diff
@@ -149,14 +155,29 @@ func (vs *Server) SubscribeCommitteeSubnets(ctx context.Context, req *ethpb.Comm
+func (vs *Server) proposeAtt(
+	ctx context.Context,
+	att ethpb.Att,
+	committee []primitives.ValidatorIndex, // required post-Electra
```
I don't understand as to why `committee` is needed here. 
`ProposeAttestationElectra` calls this function and is already passing in `SingleAttestation`
`BroadcastAttestation` doesn't require committee

potuz replied:
you need the committee to convert the attestation to an electra attestation to send the feed... perhaps that can be taken away

terencechain replied:
That's what I thought initially, but after looking closer, the feed takes `ethpb.Att`, which a single attestation already satisfies. It should theoretically work if I just pass in `SingleAttestation` to the feed, unless there's something on the receiver side that specifically requires electra attestation and not single attestation. In that case, I would argue that the receiver should fix that.

Example, this will build:
```
func (vs *Server) proposeAtt(
	ctx context.Context,
	att ethpb.Att,
	committeeIndex primitives.CommitteeIndex,
) (*ethpb.AttestResponse, error) {

	vs.OperationNotifier.OperationFeed().Send(&feed.Event{
		Type: operation.UnaggregatedAttReceived,
		Data: &operation.UnAggregatedAttReceivedData{
			Attestation: att,
		},
	})

	subnet := helpers.ComputeSubnetFromCommitteeAndSlot(uint64(len(vals)), committeeIndex, att.GetData().Slot)
	if err := vs.P2P.BroadcastAttestation(ctx, subnet, att); err != nil {
		return nil, status.Errorf(codes.Internal, "Could not broadcast attestation: %v", err)
	}
}
```

terencechain replied:
Discussed offline, we'll take the internal type to track a bit field such that we wont have to be passing committee around

terencechain commented on proto/prysm/v1alpha1/attestation.proto:
```diff
@@ -126,6 +126,25 @@ message AttestationElectra {
     bytes committee_bits = 4 [(ethereum.eth.ext.ssz_size) = "committee_bits.size", (ethereum.eth.ext.cast_type) = "committee_bits.type"];
 }
 
+message SignedAggregateAttestationAndProofSingle {
+    // The aggregated attestation and selection proof itself.
+    AggregateAttestationAndProofSingle message = 1;
+
+    // 96 byte BLS aggregate signature signed by the aggregator over the message.
+    bytes signature = 2 [(ethereum.eth.ext.ssz_size) = "96"];
+}
+
+message AggregateAttestationAndProofSingle {
+    // The aggregator index that submitted this aggregated attestation and proof.
+    uint64 aggregator_index = 1  [(ethereum.eth.ext.cast_type) = "github.com/prysmaticlabs/prysm/v5/consensus-types/primitives.ValidatorIndex"];
+
+    // The aggregated attestation that was submitted.
+    SingleAttestation aggregate = 3;
+
+    // 96 byte selection proof signed by the aggregator, which is the signature of the slot to aggregate.
```
My understanding is that this is required because of the `savePendingAtt` API. It seems annoying to define new proto types just to satisfy an API for the cache `savePendingAtt`. I wonder if there's something we can do to fix the cache for the API; probably a subsequent PR is okay

rkapka replied:
The pending atts queue would have to be redesigned so that you don't have to save `AggregateAttestationAndProof` objects. We currently construct dummy objects just to satisfy the interface.

rkapka commented on beacon-chain/core/helpers/attestation.go:
```diff
@@ -32,7 +32,7 @@ func ValidateNilAttestation(attestation ethpb.Att) error {
 		return errors.New("attestation's target can't be nil")
 	}
-	if attestation.GetAggregationBits() == nil {
+	if !attestation.IsSingle() && attestation.GetAggregationBits() == nil {
```
I will add this comment to my follow-up list

rkapka commented on beacon-chain/core/helpers/attestation_test.go:
```diff
@@ -308,6 +308,16 @@ func TestValidateNilAttestation(t *testing.T) {
 			},
 			errString: "",
 		},
+		{
```
Yes, but this is because there is no additional logic for single attestations. I can still add these tests if you feel strongly about it, but I don't see that much value in copy-pasting tests for different types when the logic is exactly the same

potuz commented on beacon-chain/sync/pending_attestations_queue.go:
```diff
@@ -91,86 +91,127 @@ func (s *Service) processPendingAtts(ctx context.Context) error {
+	if att.Version() >= version.Electra {
+		var ok bool
+		singleAtt, ok = att.(*ethpb.SingleAttestation)
+		if !ok {
```
I have a question on this path. According to the Electra spec `AggregateAndProof` [here](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/validator.md#aggregateandproof) contains an `Attestation` object inside which is never a `SingleAttestation` according to [this definition](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#attestation) in the Electra spec

potuz reviewed (commented):
Have a concern in the pending attestation queue

potuz commented on beacon-chain/sync/validate_beacon_attestation.go:
```diff
@@ -71,14 +70,6 @@ func (s *Service) validateCommitteeIndexBeaconAttestation(ctx context.Context, p
-		Data: &operation.UnAggregatedAttReceivedData{
-			Attestation: att,
-		},
-	})
 
```
Are there consumers of this feed? is this a breaking change? If so we may want to keep the deneb path unaltered by broadcasting if the version is pre-Elecrta

potuz replied:
Ahh nevermind this broadcast has moved to later.

potuz commented on beacon-chain/sync/validate_beacon_attestation.go:
```diff
@@ -115,17 +99,19 @@ func (s *Service) validateCommitteeIndexBeaconAttestation(ctx context.Context, p
+				return pubsub.ValidationIgnore, fmt.Errorf("attestation has wrong type (expected %T, got %T)", &eth.SingleAttestation{}, att)
 			}
-			s.savePendingAtt(&eth.SignedAggregateAttestationAndProofElectra{Message: &eth.AggregateAttestationAndProofElectra{Aggregate: a}})
+			s.savePendingAtt(&eth.SignedAggregateAttestationAndProofSingle{Message: &eth.AggregateAttestationAndProofSingle{Aggregate: a}})
```
Document this hack please

potuz commented on beacon-chain/sync/validate_beacon_attestation.go:
```diff
@@ -115,17 +99,19 @@ func (s *Service) validateCommitteeIndexBeaconAttestation(ctx context.Context, p
+
 	// Verify the block being voted and the processed state is in beaconDB and the block has passed validation if it's in the beaconDB.
 	blockRoot := bytesutil.ToBytes32(data.BeaconBlockRoot)
 	if !s.hasBlockAndState(ctx, blockRoot) {
```
Perhaps this whole block can be extracted from this function

potuz commented on beacon-chain/sync/validate_beacon_attestation.go:
```diff
@@ -227,78 +241,88 @@ func (s *Service) validateCommitteeIndexAndCount(
+		if !ok {
+			return pubsub.ValidationIgnore, fmt.Errorf("attestation has wrong type (expected %T, got %T)", &eth.SingleAttestation{}, a)
+		}
+		result, err := validateAttestingIndex(ctx, singleAtt.AttesterIndex, committee)
```
```suggestion
		return validateAttestingIndex(ctx, singleAtt.AttesterIndex, committee)
```
And then remove the `else` branch

potuz reviewed (commented):
Another quick pass, logic looks good to me except some nasty hacks to deal with the pending queue
